Upgrades from opam 2.1.0~beta4 to 2.1.0~rc2 remove the opam config file

```
FROM ocaml/opam
RUN opam --version # 2.0.8
RUN opam switch
RUN opam list
RUN bash -c 'echo | sh <(curl -sL https://raw.githubusercontent.com/ocaml/opam/2.1.0-rc/shell/install.sh) --dev'
RUN opam --version # 2.1.0~beta4
RUN opam switch
RUN opam list
RUN bash -c 'echo | sh <(curl -sL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh) --dev'
RUN opam --version # 2.1.0~rc2
RUN opam switch # ????
RUN opam list # ????
```
Everything goes well until the 4 last commands:
```
Step 9/12 : RUN bash -c 'echo | sh <(curl -sL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh) --dev'
 ---> Running in 25d672d35b24
## Downloading opam 2.1.0~rc2 for linux on x86_64...
## Downloaded.
## Where should it be installed ? [/usr/bin] Write access to /usr/bin required, using 'sudo'.
Command: mv /usr/bin/opam /usr/bin/opam.2.1.0~beta4
## /usr/bin/opam backed up as opam.2.1.0~beta4
## Backing up /home/opam/.opam to .opam.2.1.0~beta4 (this may take a while)
## /home/opam/.opam backed up as .opam.2.1.0~beta4
Write access to /usr/bin required, using 'sudo'.
Command: install -m 755 /tmp/opam-2.1.0-rc2-x86_64-linux /usr/bin/opam
## opam 2.1.0~rc2 installed to /usr/bin
## Converting the opam root format & updating
Configuring from /home/opam/.opamrc and then from built-in defaults.
Checking for available remotes: [ERROR] In /home/opam/.opam/config:
        unsupported or missing file format version; should be 2.0 or older [skipped]

rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.

This version of opam requires an update to the layout of /home/opam/.opam from version 2.0 to version 2.1, which can't be reverted.
You may want to back it up before going further.

Continue? [Y/n] y
Format upgrade done.

<><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
[default] no changes from file:///home/opam/opam-repository
## Run this script again with '--restore 2.1.0~beta4' to revert.
Removing intermediate container 25d672d35b24
 ---> b02b09b4e36c
Step 10/12 : RUN opam --version # 2.1.0~rc2
 ---> Running in dba4eb5dc8b9
2.1.0~rc2
Removing intermediate container dba4eb5dc8b9
 ---> 102e94473826
Step 11/12 : RUN opam switch # ????
 ---> Running in a0c5d1ad38e7
#  switch  compiler  description
Removing intermediate container a0c5d1ad38e7
 ---> 40cf093b1ec3
Step 12/12 : RUN opam list # ????
 ---> Running in a04b5dd02d7f
# No matches found
```
Looking inside the `~/.opam` directory reveals a mostly unchanged opam root, however `~/.opam/config` does not have any switches:
```diff
opam@92a8720d34dd:~$ diff -ru .opam.2.1.0~beta4/ .opam
diff -ru .opam.2.1.0~beta4/config .opam/config
--- .opam.2.1.0~beta4/config	2021-07-09 14:37:34.000000000 +0100
+++ .opam/config	2021-07-09 14:48:14.688880274 +0100
@@ -1,17 +1,43 @@
-opam-version: "2.1"
-repositories: "default"
-installed-switches: "4.12"
-switch: "4.12"
-jobs: 31
+opam-version: "2.0"
+opam-root-version: "2.1"
 download-jobs: 3
 eval-variables: [
-  sys-ocaml-version
-  ["ocamlc" "-vnum"]
-  "OCaml version present on your system independently of opam, if any"
+  [
+    sys-ocaml-version
+    ["ocamlc" "-vnum"]
+    "OCaml version present on your system independently of opam, if any"
+  ]
+  [
+    sys-ocaml-arch
+    [
+      "sh"
+      "-c"
+      "ocamlc -config 2>/dev/null | tr -d '\\r' | grep '^architecture: ' | sed -e 's/.*: //' -e 's/i386/i686/' -e 's/amd64/x86_64/'"
+    ]
+    "Target architecture of the OCaml compiler present on your system"
+  ]
+  [
+    sys-ocaml-cc
+    [
+      "sh"
+      "-c"
+      "ocamlc -config 2>/dev/null | tr -d '\\r' | grep '^ccomp_type: ' | sed -e 's/.*: //'"
+    ]
+    "Host C Compiler type of the OCaml compiler present on your system"
+  ]
+  [
+    sys-ocaml-libc
+    [
+      "sh"
+      "-c"
+      "ocamlc -config 2>/dev/null | tr -d '\\r' | grep '^os_type: ' | sed -e 's/.*: //' -e 's/Win32/msvc/' -e '/^msvc$/!s/.*/libc/'"
+    ]
+    "Host C Runtime Library type of the OCaml compiler present on your system"
+  ]
 ]
-default-compiler: [
-  "ocaml-system" {>= "4.02.3"}
-  "ocaml-base-compiler"
+default-compiler: ["ocaml-system" "ocaml-base-compiler"]
+default-invariant: [
+  "ocaml" {>= "4.05.0"}
 ]
 depext: true
 depext-run-installs: true
diff -ru .opam.2.1.0~beta4/opam-init/env_hook.sh .opam/opam-init/env_hook.sh
--- .opam.2.1.0~beta4/opam-init/env_hook.sh	2021-07-09 14:37:34.000000000 +0100
+++ .opam/opam-init/env_hook.sh	2021-07-09 14:48:14.688880274 +0100
@@ -1,7 +1,7 @@
 OPAMNOENVNOTICE=true; export OPAMNOENVNOTICE;
 _opam_env_hook() {
  local previous_exit_status=$?;
- eval $(opam env --shell=bash --readonly 2> /dev/null);
+ eval $(opam env --shell=bash --readonly 2> /dev/null <&- );
  return $previous_exit_status;
 };
 if ! [[ "$PROMPT_COMMAND" =~ _opam_env_hook ]]; then
diff -ru .opam.2.1.0~beta4/opam-init/env_hook.zsh .opam/opam-init/env_hook.zsh
--- .opam.2.1.0~beta4/opam-init/env_hook.zsh	2021-07-09 14:37:34.000000000 +0100
+++ .opam/opam-init/env_hook.zsh	2021-07-09 14:48:14.688880274 +0100
@@ -1,6 +1,6 @@
 OPAMNOENVNOTICE=true; export OPAMNOENVNOTICE;
 _opam_env_hook() {
-    eval $(opam env --shell=zsh --readonly 2> /dev/null);
+    eval $(opam env --shell=zsh --readonly 2> /dev/null <&-);
 }
 typeset -ag precmd_functions;
 if [[ -z ${precmd_functions[(r)_opam_env_hook]} ]]; then
diff -ru .opam.2.1.0~beta4/opam-init/hooks/sandbox.sh .opam/opam-init/hooks/sandbox.sh
--- .opam.2.1.0~beta4/opam-init/hooks/sandbox.sh	2021-07-09 14:37:34.000000000 +0100
+++ .opam/opam-init/hooks/sandbox.sh	2021-07-09 14:48:14.684880247 +0100
@@ -11,10 +11,12 @@
     exit 10
 fi
 
-ARGS=(--unshare-net --new-session)
+# --new-session requires bubblewrap 0.1.7
+# --die-with-parent requires bubblewrap 0.1.8
+ARGS=(--unshare-net --new-session --die-with-parent)
 ARGS=("${ARGS[@]}" --proc /proc --dev /dev)
-ARGS=("${ARGS[@]}" --bind "${TMPDIR:-/tmp}" /tmp)
-ARGS=("${ARGS[@]}" --setenv TMPDIR /tmp --setenv TMP /tmp --setenv TEMPDIR /tmp --setenv TEMP /tmp)
+ARGS=("${ARGS[@]}" --setenv TMPDIR /opam-tmp --setenv TMP /opam-tmp --setenv TEMPDIR /opam-tmp --setenv TEMP /opam-tmp)
+ARGS=("${ARGS[@]}" --tmpfs /opam-tmp)
 ARGS=("${ARGS[@]}" --tmpfs /run)
 
 add_mount() {
@@ -55,7 +57,7 @@
 # use OPAM_USER_PATH_RO variable to add them
 # the OPAM_USER_PATH_RO format is the same as PATH
 # ie: export OPAM_USER_PATH_RO=/nix/store:/rw/usrlocal
-add_sys_mounts /usr /bin /lib /lib32 /lib64 /etc /opt /home /var
+add_sys_mounts /usr /bin /lib /lib32 /lib64 /etc /opt /home /var /tmp
 
 # C compilers using `ccache` will write to a shared cache directory
 # that remain writeable. ccache seems widespread in some Fedora systems.
@@ -76,15 +78,20 @@
 }
 
 add_dune_cache_mount() {
-  u_cache=${XDG_CACHE_HOME:-$HOME/.cache}
-  u_dune_cache=$u_cache/dune
-  cache=$(readlink -m "$u_cache")
-  dune_cache=$cache/dune
-  dune_cache=$(readlink -m "$u_dune_cache")
+  local u_cache=${XDG_CACHE_HOME:-$HOME/.cache}
+  local u_dune_cache=$u_cache/dune
+  local cache=$(readlink -m "$u_cache")
+  local dune_cache=$cache/dune
+  local dune_cache=$(readlink -m "$u_dune_cache")
   mkdir -p "${dune_cache}"
   add_mount rw "$u_dune_cache" "$dune_cache"
 }
 
+# mount unusual path in ro
+if  [ -n "${OPAM_USER_PATH_RO-}" ]; then
+   add_mounts ro $(echo "${OPAM_USER_PATH_RO}" | sed 's|:| |g')
+fi
+
 # When using opam variable that must be defined at action time, add them also
 # at init check in OpamAuxCommands.check_and_revert_sandboxing (like
 # OPAM_SWITCH_PREFIX).
@@ -92,32 +99,20 @@
 COMMAND="$1"; shift
 case "$COMMAND" in
     build)
-        # mount unusual path in ro
-        if  [ -n "${OPAM_USER_PATH_RO-}" ]; then
-           add_mounts ro $(echo "${OPAM_USER_PATH_RO}" | sed 's|:| |g')
-        fi
         add_mounts ro "$OPAM_SWITCH_PREFIX"
         add_mounts rw "$PWD"
         add_ccache_mount
         add_dune_cache_mount
         ;;
     install)
-        # mount unusual path in ro
-        if  [ -n "${OPAM_USER_PATH_RO-}" ]; then
-           add_mounts ro  $(echo "${OPAM_USER_PATH_RO}" | sed 's|:| |g')
-        fi
         add_mounts rw "$OPAM_SWITCH_PREFIX"
         add_mounts ro "$OPAM_SWITCH_PREFIX/.opam-switch"
         add_mounts rw "$PWD"
         ;;
     remove)
-        # mount unusual path in ro
-        if  [ -n "${OPAM_USER_PATH_RO-}" ]; then
-           add_mounts ro $(echo "${OPAM_USER_PATH_RO}" | sed 's|:| |g')
-        fi
         add_mounts rw "$OPAM_SWITCH_PREFIX"
         add_mounts ro "$OPAM_SWITCH_PREFIX/.opam-switch"
-        if [ "X${PWD#$OPAM_SWITCH_PREFIX}/.opam-switch/" != "X${PWD}" ]; then
+        if [ "X${PWD#$OPAM_SWITCH_PREFIX/.opam-switch/}" != "X${PWD}" ]; then
           add_mounts rw "$PWD"
         fi
         ;;
@@ -127,4 +122,5 @@
 esac
 
 # Note: we assume $1 can be trusted, see https://github.com/projectatomic/bubblewrap/issues/259
+# As of now we are compatible up to 0.1.8, '--' can be added here when we require >= 0.3.0
 exec bwrap "${ARGS[@]}" "$@"
diff -ru .opam.2.1.0~beta4/opam-init/hooks/sandbox.sh.hash .opam/opam-init/hooks/sandbox.sh.hash
--- .opam.2.1.0~beta4/opam-init/hooks/sandbox.sh.hash	2021-07-09 14:37:34.000000000 +0100
+++ .opam/opam-init/hooks/sandbox.sh.hash	2021-07-09 14:48:14.684880247 +0100
@@ -1 +1 @@
-md5=c51ece9376c419167196f4438f0babcd
\ No newline at end of file
+md5=0ff6dc2bfbe090e3944ce2674aa21357
\ No newline at end of file
Only in .opam.2.1.0~beta4/plugins/bin: opam-depext
Binary files .opam.2.1.0~beta4/repo/default.tar.gz and .opam/repo/default.tar.gz differ
diff -ru .opam.2.1.0~beta4/repo/repos-config .opam/repo/repos-config
--- .opam.2.1.0~beta4/repo/repos-config	2021-07-09 14:39:03.000000000 +0100
+++ .opam/repo/repos-config	2021-07-09 14:49:27.877377522 +0100
@@ -1 +1,2 @@
+opam-version: "2.0"
 repositories: "default" {"file:///home/opam/opam-repository"}
Only in .opam/repo: state-21F1960C.cache
Only in .opam.2.1.0~beta4/repo: state.cache
```
Hit by @NathanReb

--- a/CHANGES
+++ b/CHANGES
@@ -6,6 +6,8 @@ are not marked). Those prefixed with "(+)" are new command/option (since
 2.1.0:
 * Set DEBIAN_FRONTEND=noninteractive for unsafe-yes confirmation level
   [#4735 @dra27 - partially fix #4731]
+* Fix 2.1~alpha2 to 2.1 format upgrade with reinit [#4750 @rjbou - fix #4748]
+* Fix bypass-check handling on reinit [#4750 @rjbou]
 
 2.1.0~rc2:
 * Remove OPAMZ3DEBUG evironment variable [#4720 @rjbou - fix #4717]

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -371,10 +371,18 @@ let init cli =
           get_init_config ~no_sandboxing
             ~no_default_config_file:no_config_file ~add_config_file:config_file
         in
+        let config =
+          match OpamStateConfig.load ~lock_kind:`Lock_write root with
+          | Some c -> c
+          | exception (OpamPp.Bad_version _ as e) ->
+            OpamFormatUpgrade.hard_upgrade_from_2_1_intermediates root;
+            raise e
+          | None -> OpamFile.Config.empty
+        in
         OpamClient.reinit ~init_config ~interactive ~dot_profile
           ?update_config ?env_hook ?completion ~inplace
           ~check_sandbox:(not no_sandboxing)
-          (OpamStateConfig.safe_load ~lock_kind:`Lock_write root) shell;
+          config shell
       else
         (if not interactive &&
             update_config <> Some true && completion <> Some true && env_hook <> Some true then

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -380,7 +380,7 @@ let init cli =
           | None -> OpamFile.Config.empty
         in
         OpamClient.reinit ~init_config ~interactive ~dot_profile
-          ?update_config ?env_hook ?completion ~inplace
+          ?update_config ?env_hook ?completion ~inplace ~bypass_checks
           ~check_sandbox:(not no_sandboxing)
           config shell
       else

--- a/src/client/opamConfigCommand.ml
+++ b/src/client/opamConfigCommand.ml
@@ -573,11 +573,15 @@ let with_switch:
   | None ->
     let switch = OpamStateConfig.get_switch () in
     let switch_config =
-      try OpamStateConfig.Switch.safe_load ~lock_kind gt switch
-      with OpamPp.Bad_version _ as e ->
-        OpamFormatUpgrade.hard_upgrade_from_2_1_intermediates
-          ~global_lock:gt.global_lock gt.root;
-        raise e
+      if lock_kind = `Lock_write then
+        match OpamStateConfig.Switch.read_opt ~lock_kind gt switch with
+        | Some c -> c
+        | exception (OpamPp.Bad_version _ as e) ->
+          OpamFormatUpgrade.hard_upgrade_from_2_1_intermediates gt.root;
+          raise e
+        | None -> OpamFile.Switch_config.empty
+      else
+        OpamStateConfig.Switch.safe_load ~lock_kind gt switch
     in
     let lock_file = OpamPath.Switch.lock gt.root switch in
     if switch_config = OpamFile.Switch_config.empty then

--- a/src/state/opamGlobalState.ml
+++ b/src/state/opamGlobalState.ml
@@ -60,7 +60,12 @@ let load lock_kind =
     OpamConsole.error_and_exit `Configuration_error
       "Opam has not been initialised, please run `opam init'";
   let config_lock = OpamFilename.flock lock_kind (OpamPath.config_lock root) in
-  let config = load_config lock_kind global_lock root in
+  let config =
+    try load_config lock_kind global_lock root
+    with OpamFormatUpgrade.Upgrade_done _ as e ->
+      OpamSystem.funlock config_lock;
+      raise e
+  in
   if OpamStateConfig.is_newer config && lock_kind <> `Lock_write then
     log "root version (%s) is greater than running binary's (%s); \
          load with best-effort (read-only)"

--- a/src/state/opamStateConfig.ml
+++ b/src/state/opamStateConfig.ml
@@ -232,9 +232,13 @@ let load_config_root ?lock_kind readf opamroot =
     (OpamFile.Config.raw_root_version f)
     readf f
 
+let safe read read' default =
+  let safe r f = OpamStd.Option.default default @@ r f in
+  safe read, safe read'
+
 let safe_load ?lock_kind opamroot =
   load_config_root ?lock_kind
-    OpamFile.Config.(safe_read, BestEffort.safe_read) opamroot
+    OpamFile.Config.(safe read_opt BestEffort.read_opt empty) opamroot
 
 let load ?lock_kind opamroot =
   load_config_root ?lock_kind
@@ -250,7 +254,7 @@ module Switch = struct
   let safe_load_t ?lock_kind root switch =
     let config = safe_load ~lock_kind:`Lock_read root in
     load_raw ?lock_kind root config
-      OpamFile.Switch_config.(safe_read, BestEffort.safe_read)
+      OpamFile.Switch_config.(safe read_opt BestEffort.read_opt empty)
       switch
 
   let load ?lock_kind gt readf switch =
@@ -258,7 +262,7 @@ module Switch = struct
 
   let safe_load ?lock_kind gt switch =
     load ?lock_kind gt
-      OpamFile.Switch_config.(safe_read, BestEffort.safe_read)
+      OpamFile.Switch_config.(safe read_opt BestEffort.read_opt empty)
       switch
 
   let read_opt ?lock_kind gt switch =
@@ -268,7 +272,7 @@ module Switch = struct
 
   let safe_read_selections ?lock_kind gt switch =
     load_if_possible ?lock_kind gt
-      OpamFile.SwitchSelections.(safe_read, BestEffort.safe_read)
+      OpamFile.SwitchSelections.(safe read_opt BestEffort.read_opt empty)
       (OpamPath.Switch.selections gt.root switch)
 
 end
@@ -277,7 +281,7 @@ end
 module Repos = struct
   let safe_read ?lock_kind gt =
     load_if_possible ?lock_kind gt
-      OpamFile.Repos_config.(safe_read, BestEffort.safe_read)
+      OpamFile.Repos_config.(safe read_opt BestEffort.read_opt empty)
       (OpamPath.repos_config gt.root)
 end
 

--- a/tests/reftests/opamroot-versions.test
+++ b/tests/reftests/opamroot-versions.test
@@ -933,6 +933,60 @@ compiler: ["i-am-sys-compiler.2"]
 installed: ["i-am-package.2" "i-am-sys-compiler.2"]
 opam-version: "2.0"
 roots: ["i-am-package.2" "i-am-sys-compiler.2"]
+### :V:1:e: reinit from 2.0
+### ocaml generate.ml 2.0
+### opam init --reinit --bypass-checks --no-setup
+No configuration file found, using built-in defaults.
+Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+[WARNING] Removing global switch 'this-internal-error' as it no longer exists
+FMT_UPG                         Light config upgrade, from 2.0 to 2.1
+This version of opam requires an update to the layout of ${BASEDIR}/OPAM from version 2.0 to version 2.1, which can't be reverted.
+You may want to back it up before going further.
+
+Continue? [Y/n] y
+Format upgrade done.
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+
+<><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
+[default] no changes from file://${BASEDIR}/default
+### opam switch
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+#   switch                                     compiler             description
+->  ${BASEDIR}  i-am-sys-compiler.2  local switch
+    default                                                         default switch
+    sw-comp                                    i-am-compiler.2      switch with compiler
+    sw-sys-comp                                                     switch with system compiler
+
+[NOTE] Current switch has been selected based on the current directory.
+       The current global system switch is sw-sys-comp.
+### opam list
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+STATE                           LOAD-SWITCH-STATE @ ${BASEDIR}
+STATE                           Switch state loaded in 0.000s
+# Packages matching: installed
+# Name            # Installed # Synopsis
+i-am-package      2           One-line description
+i-am-sys-compiler 2           One-line description
+### opam-cat $OPAMROOT/config
+default-compiler: ["i-am-sys-compiler" "i-am-compiler"]
+default-invariant: ["ocaml" {>= "4.05.0"}]
+depext: true
+depext-cannot-install: false
+depext-run-installs: true
+download-jobs: 3
+eval-variables: [sys-comp-version ["sh" "-c" "echo $OPAMSYSCOMP"] "comp version"]
+installed-switches: ["sw-sys-comp" "sw-comp" "default"]
+opam-root-version: "2.1"
+opam-version: "2.0"
+repositories: "default"
+switch: "sw-sys-comp"
+wrap-build-commands: ["%{hooks}%/sandbox.sh" "build"] {os = "linux" | os = "macos"}
+wrap-install-commands: ["%{hooks}%/sandbox.sh" "install"] {os = "linux" | os = "macos"}
+wrap-remove-commands: ["%{hooks}%/sandbox.sh" "remove"] {os = "linux" | os = "macos"}
 ### rm -rf _opam
 ### :V:2:a: From 2.1~alpha root, global
 ### ocaml generate.ml 2.1~alpha
@@ -1215,6 +1269,65 @@ compiler: ["i-am-sys-compiler.2"]
 installed: ["i-am-package.2" "i-am-sys-compiler.2"]
 opam-version: "2.0"
 roots: ["i-am-package.2" "i-am-sys-compiler.2"]
+### :V:2:e: reinit from 2.1~alpha
+### ocaml generate.ml 2.1~alpha
+### opam init --reinit --bypass-checks --no-setup
+No configuration file found, using built-in defaults.
+Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+[WARNING] Removing global switch 'this-internal-error' as it no longer exists
+FMT_UPG                         Hard config upgrade, from 2.1~alpha to 2.1
+This version of opam requires an update to the layout of ${BASEDIR}/OPAM from version 2.1~alpha to version 2.1, which can't be reverted.
+You may want to back it up before going further.
+
+Perform the update and continue? [Y/n] y
+Format upgrade done.
+
+<><> Rerunning init and update ><><><><><><><><><><><><><><><><><><><><><><><><>
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+
+<><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
+[default] no changes from file://${BASEDIR}/default
+Update done, please now retry your command.
+# Return code 10 #
+### opam switch
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+#   switch                                     compiler             description
+->  ${BASEDIR}  i-am-sys-compiler.2  local switch
+    default                                                         default switch
+    sw-comp                                    i-am-compiler.2      switch with compiler
+    sw-sys-comp                                                     switch with system compiler
+
+[NOTE] Current switch has been selected based on the current directory.
+       The current global system switch is sw-sys-comp.
+### opam list
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+STATE                           LOAD-SWITCH-STATE @ ${BASEDIR}
+STATE                           Switch state loaded in 0.000s
+# Packages matching: installed
+# Name            # Installed # Synopsis
+i-am-package      2           One-line description
+i-am-sys-compiler 2           One-line description
+### opam-cat $OPAMROOT/config
+default-compiler: ["i-am-sys-compiler" "i-am-compiler"]
+default-invariant: ["ocaml" {>= "4.05.0"}]
+depext: true
+depext-cannot-install: false
+depext-run-installs: true
+download-jobs: 3
+eval-variables: [sys-comp-version ["sh" "-c" "echo $OPAMSYSCOMP"] "comp version"]
+installed-switches: ["sw-sys-comp" "sw-comp" "default"]
+opam-root-version: "2.1"
+opam-version: "2.0"
+repositories: "default"
+switch: "sw-sys-comp"
+wrap-build-commands: ["%{hooks}%/sandbox.sh" "build"] {os = "linux" | os = "macos"}
+wrap-install-commands: ["%{hooks}%/sandbox.sh" "install"] {os = "linux" | os = "macos"}
+wrap-remove-commands: ["%{hooks}%/sandbox.sh" "remove"] {os = "linux" | os = "macos"}
 ### rm -rf _opam
 ### :V:3:a: From 2.1~alpha2 root, global
 ### ocaml generate.ml 2.1~alpha2
@@ -1504,6 +1617,65 @@ compiler: ["i-am-sys-compiler.2"]
 installed: ["i-am-package.2" "i-am-sys-compiler.2"]
 opam-version: "2.0"
 roots: ["i-am-package.2" "i-am-sys-compiler.2"]
+### :V:3:e: reinit from 2.1~alpha2
+### ocaml generate.ml 2.1~alpha2
+### opam init --reinit --bypass-checks --no-setup
+No configuration file found, using built-in defaults.
+FMT_UPG                         Intermediate opam root detected, launch hard upgrade
+FMT_UPG                         Downgrade config opam-version to fix up
+[WARNING] Removing global switch 'this-internal-error' as it no longer exists
+FMT_UPG                         Hard config upgrade, from 2.1~alpha2 to 2.1
+This version of opam requires an update to the layout of ${BASEDIR}/OPAM from version 2.1~alpha2 to version 2.1, which can't be reverted.
+You may want to back it up before going further.
+
+Perform the update and continue? [Y/n] y
+Format upgrade done.
+
+<><> Rerunning init and update ><><><><><><><><><><><><><><><><><><><><><><><><>
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+
+<><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
+[default] no changes from file://${BASEDIR}/default
+Update done, please now retry your command.
+# Return code 10 #
+### opam switch
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+#   switch                                     compiler             description
+->  ${BASEDIR}  i-am-sys-compiler.2  local switch
+    default                                                         default switch
+    sw-comp                                    i-am-compiler.2      switch with compiler
+    sw-sys-comp                                                     switch with system compiler
+
+[NOTE] Current switch has been selected based on the current directory.
+       The current global system switch is sw-sys-comp.
+### opam list
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+STATE                           LOAD-SWITCH-STATE @ ${BASEDIR}
+STATE                           Switch state loaded in 0.000s
+# Packages matching: installed
+# Name            # Installed # Synopsis
+i-am-package      2           One-line description
+i-am-sys-compiler 2           One-line description
+### opam-cat $OPAMROOT/config
+default-compiler: ["i-am-sys-compiler" "i-am-compiler"]
+default-invariant: ["ocaml" {>= "4.05.0"}]
+depext: true
+depext-cannot-install: false
+depext-run-installs: true
+download-jobs: 3
+eval-variables: [sys-comp-version ["sh" "-c" "echo $OPAMSYSCOMP"] "comp version"]
+installed-switches: ["sw-sys-comp" "sw-comp" "default"]
+opam-root-version: "2.1"
+opam-version: "2.0"
+repositories: "default"
+switch: "sw-sys-comp"
+wrap-build-commands: ["%{hooks}%/sandbox.sh" "build"] {os = "linux" | os = "macos"}
+wrap-install-commands: ["%{hooks}%/sandbox.sh" "install"] {os = "linux" | os = "macos"}
+wrap-remove-commands: ["%{hooks}%/sandbox.sh" "remove"] {os = "linux" | os = "macos"}
 ### rm -rf _opam
 ### :V:4:a: From 2.1~rc root, global
 ### ocaml generate.ml 2.1~rc
@@ -1755,6 +1927,60 @@ compiler: ["i-am-sys-compiler.2"]
 installed: ["i-am-package.2" "i-am-sys-compiler.2"]
 opam-version: "2.0"
 roots: ["i-am-package.2" "i-am-sys-compiler.2"]
+### :V:4:e: reinit from 2.1~rc
+### ocaml generate.ml 2.1~rc
+### opam init --reinit --bypass-checks --no-setup
+No configuration file found, using built-in defaults.
+Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+[WARNING] Removing global switch 'this-internal-error' as it no longer exists
+FMT_UPG                         Light config upgrade, from 2.1~rc to 2.1
+This version of opam requires an update to the layout of ${BASEDIR}/OPAM from version 2.1~rc to version 2.1, which can't be reverted.
+You may want to back it up before going further.
+
+Continue? [Y/n] y
+Format upgrade done.
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+
+<><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
+[default] no changes from file://${BASEDIR}/default
+### opam switch
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+#   switch                                     compiler             description
+->  ${BASEDIR}  i-am-sys-compiler.2  local switch
+    default                                                         default switch
+    sw-comp                                    i-am-compiler.2      switch with compiler
+    sw-sys-comp                                                     switch with system compiler
+
+[NOTE] Current switch has been selected based on the current directory.
+       The current global system switch is sw-sys-comp.
+### opam list
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+STATE                           LOAD-SWITCH-STATE @ ${BASEDIR}
+STATE                           Switch state loaded in 0.000s
+# Packages matching: installed
+# Name            # Installed # Synopsis
+i-am-package      2           One-line description
+i-am-sys-compiler 2           One-line description
+### opam-cat $OPAMROOT/config
+default-compiler: ["i-am-sys-compiler" "i-am-compiler"]
+default-invariant: ["i-am-sys-compiler"]
+depext: true
+depext-cannot-install: false
+depext-run-installs: true
+download-jobs: 3
+eval-variables: [sys-comp-version ["sh" "-c" "echo $OPAMSYSCOMP"] "comp version"]
+installed-switches: ["sw-sys-comp" "sw-comp" "default"]
+opam-root-version: "2.1"
+opam-version: "2.0"
+repositories: "default"
+switch: "sw-sys-comp"
+wrap-build-commands: ["%{hooks}%/sandbox.sh" "build"] {os = "linux" | os = "macos"}
+wrap-install-commands: ["%{hooks}%/sandbox.sh" "install"] {os = "linux" | os = "macos"}
+wrap-remove-commands: ["%{hooks}%/sandbox.sh" "remove"] {os = "linux" | os = "macos"}
 ### rm -rf _opam
 ### :V:5:a: From 2.1 root, global
 ### ocaml generate.ml 2.1
@@ -1954,3 +2180,51 @@ compiler: ["i-am-sys-compiler.2"]
 installed: ["i-am-package.2" "i-am-sys-compiler.2"]
 opam-version: "2.0"
 roots: ["i-am-package.2" "i-am-sys-compiler.2"]
+### :V:5:e: reinit from 2.1
+### ocaml generate.ml 2.1
+### opam init --reinit --bypass-checks --no-setup
+No configuration file found, using built-in defaults.
+Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+
+<><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
+[default] no changes from file://${BASEDIR}/default
+### opam-cat $OPAMROOT/config
+default-compiler: ["i-am-sys-compiler" "i-am-compiler"]
+default-invariant: ["i-am-sys-compiler"]
+depext: true
+depext-cannot-install: false
+depext-run-installs: true
+download-jobs: 3
+eval-variables: [sys-comp-version ["sh" "-c" "echo $OPAMSYSCOMP"] "comp version"]
+installed-switches: ["sw-sys-comp" "sw-comp" "default" "${BASEDIR}/why-did-you-delete-me" "this-internal-error"]
+opam-root-version: "2.1"
+opam-version: "2.0"
+repositories: "default"
+switch: "sw-sys-comp"
+wrap-build-commands: ["%{hooks}%/sandbox.sh" "build"] {os = "linux" | os = "macos"}
+wrap-install-commands: ["%{hooks}%/sandbox.sh" "install"] {os = "linux" | os = "macos"}
+wrap-remove-commands: ["%{hooks}%/sandbox.sh" "remove"] {os = "linux" | os = "macos"}
+### opam switch
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+#   switch                                     compiler             description
+->  ${BASEDIR}  i-am-sys-compiler.2  local switch
+    default                                                         default switch
+    sw-comp                                    i-am-compiler.2      switch with compiler
+    sw-sys-comp                                                     switch with system compiler
+    this-internal-error                                             Missing config file
+
+[NOTE] Current switch has been selected based on the current directory.
+       The current global system switch is sw-sys-comp.
+### opam list
+GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
+STATE                           LOAD-SWITCH-STATE @ ${BASEDIR}
+STATE                           Switch state loaded in 0.000s
+# Packages matching: installed
+# Name            # Installed # Synopsis
+i-am-package      2           One-line description
+i-am-sys-compiler 2           One-line description

--- a/tests/reftests/opamroot-versions.test
+++ b/tests/reftests/opamroot-versions.test
@@ -937,7 +937,6 @@ roots: ["i-am-package.2" "i-am-sys-compiler.2"]
 ### ocaml generate.ml 2.0
 ### opam init --reinit --bypass-checks --no-setup
 No configuration file found, using built-in defaults.
-Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 [WARNING] Removing global switch 'this-internal-error' as it no longer exists
 FMT_UPG                         Light config upgrade, from 2.0 to 2.1
@@ -1273,7 +1272,6 @@ roots: ["i-am-package.2" "i-am-sys-compiler.2"]
 ### ocaml generate.ml 2.1~alpha
 ### opam init --reinit --bypass-checks --no-setup
 No configuration file found, using built-in defaults.
-Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 [WARNING] Removing global switch 'this-internal-error' as it no longer exists
 FMT_UPG                         Hard config upgrade, from 2.1~alpha to 2.1
@@ -1931,7 +1929,6 @@ roots: ["i-am-package.2" "i-am-sys-compiler.2"]
 ### ocaml generate.ml 2.1~rc
 ### opam init --reinit --bypass-checks --no-setup
 No configuration file found, using built-in defaults.
-Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 [WARNING] Removing global switch 'this-internal-error' as it no longer exists
 FMT_UPG                         Light config upgrade, from 2.1~rc to 2.1
@@ -2184,7 +2181,6 @@ roots: ["i-am-package.2" "i-am-sys-compiler.2"]
 ### ocaml generate.ml 2.1
 ### opam init --reinit --bypass-checks --no-setup
 No configuration file found, using built-in defaults.
-Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
 RSTATE                          Cache found

--- a/tests/reftests/opamroot-versions.test
+++ b/tests/reftests/opamroot-versions.test
@@ -950,16 +950,12 @@ RSTATE                          Cache found
 
 <><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
 [default] no changes from file://${BASEDIR}/default
-### opam switch
+### opam switch --short
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
-#   switch                                     compiler             description
-->  ${BASEDIR}  i-am-sys-compiler.2  local switch
-    default                                                         default switch
-    sw-comp                                    i-am-compiler.2      switch with compiler
-    sw-sys-comp                                                     switch with system compiler
-
-[NOTE] Current switch has been selected based on the current directory.
-       The current global system switch is sw-sys-comp.
+${BASEDIR}
+default
+sw-comp
+sw-sys-comp
 ### opam list
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
@@ -1290,16 +1286,12 @@ RSTATE                          Cache found
 [default] no changes from file://${BASEDIR}/default
 Update done, please now retry your command.
 # Return code 10 #
-### opam switch
+### opam switch --short
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
-#   switch                                     compiler             description
-->  ${BASEDIR}  i-am-sys-compiler.2  local switch
-    default                                                         default switch
-    sw-comp                                    i-am-compiler.2      switch with compiler
-    sw-sys-comp                                                     switch with system compiler
-
-[NOTE] Current switch has been selected based on the current directory.
-       The current global system switch is sw-sys-comp.
+${BASEDIR}
+default
+sw-comp
+sw-sys-comp
 ### opam list
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
@@ -1638,16 +1630,12 @@ RSTATE                          Cache found
 [default] no changes from file://${BASEDIR}/default
 Update done, please now retry your command.
 # Return code 10 #
-### opam switch
+### opam switch --short
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
-#   switch                                     compiler             description
-->  ${BASEDIR}  i-am-sys-compiler.2  local switch
-    default                                                         default switch
-    sw-comp                                    i-am-compiler.2      switch with compiler
-    sw-sys-comp                                                     switch with system compiler
-
-[NOTE] Current switch has been selected based on the current directory.
-       The current global system switch is sw-sys-comp.
+${BASEDIR}
+default
+sw-comp
+sw-sys-comp
 ### opam list
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
@@ -1942,16 +1930,12 @@ RSTATE                          Cache found
 
 <><> Updating repositories ><><><><><><><><><><><><><><><><><><><><><><><><><><>
 [default] no changes from file://${BASEDIR}/default
-### opam switch
+### opam switch --short
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
-#   switch                                     compiler             description
-->  ${BASEDIR}  i-am-sys-compiler.2  local switch
-    default                                                         default switch
-    sw-comp                                    i-am-compiler.2      switch with compiler
-    sw-sys-comp                                                     switch with system compiler
-
-[NOTE] Current switch has been selected based on the current directory.
-       The current global system switch is sw-sys-comp.
+${BASEDIR}
+default
+sw-comp
+sw-sys-comp
 ### opam list
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
@@ -2203,17 +2187,13 @@ switch: "sw-sys-comp"
 wrap-build-commands: ["%{hooks}%/sandbox.sh" "build"] {os = "linux" | os = "macos"}
 wrap-install-commands: ["%{hooks}%/sandbox.sh" "install"] {os = "linux" | os = "macos"}
 wrap-remove-commands: ["%{hooks}%/sandbox.sh" "remove"] {os = "linux" | os = "macos"}
-### opam switch
+### opam switch --short
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
-#   switch                                     compiler             description
-->  ${BASEDIR}  i-am-sys-compiler.2  local switch
-    default                                                         default switch
-    sw-comp                                    i-am-compiler.2      switch with compiler
-    sw-sys-comp                                                     switch with system compiler
-    this-internal-error                                             Missing config file
-
-[NOTE] Current switch has been selected based on the current directory.
-       The current global system switch is sw-sys-comp.
+${BASEDIR}
+default
+sw-comp
+sw-sys-comp
+this-internal-error
 ### opam list
 GSTATE                          LOAD-GLOBAL-STATE @ ${BASEDIR}/OPAM
 RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM

